### ISSUE TITLE
mod_sftp unsuccessful login count issues on AIX

### ISSUE THREAD

**Issue by ws0w** [proftpd/proftpd#693]
### What I Did

I do not use AIX by choice.

On AIX 7.1 using gcc I compiled proftpd (latest master branch from github) with mod_sftp, and configured proftpd to allow SSH keys or passwords for authentication. I did not set up the ssh keys for the id tested.

Problem 1) When I sftp to proftpd, the unsuccessful login count gets incremented twice before the user is prompted for a password. From the trace log, I can see that it actually increments the count twice before checking for the SSH key. The unsuccessful login count should not be incremented until the login has actually failed (like OpenSSH' sftp). 

Problem 2) After successful login with a password, proftpd fails to rest the unsuccessful login count because root privileges have already been revoked before the login completed and the call to AIX's loginsuccess() fails.

### What I Expected/Wanted

I expected that when I sftped to the server I would be prompted for a password. After successfully logging in using a password, I expected the user's unsuccessful login count to be 0 and other  details maintained by the OS in /etc/securoty/lastlog to be updated appropriately. This is all accomplished by a call to AIX's loginsuccess() system call.

### ProFTPD Version and Configuration
```
Compile-time Settings:
  Version: 1.3.7rc1 (git)
  Platform: AIX7 (AIX7_1_4_0) [AIX 1 00C21E024C00]
  Built: Thu Apr 26 2018 14:43:33 MDT
  Built With:
    configure  '--prefix=/opt/proftpd' '--with-modules=mod_sftp' 'CFLAGS=-DPR_TUNABLE_XFER_LOG_MODE=0664 -DAIX7 -I/home/dan/proftpd-git -I/home/dan/proftpd-git/include -maix64' 'LDFLAGS=-dn -L/opt/freeware/lib -L/opt/freeware/lib64 -maix64 -Wl,-bbigtoc'

  CFLAGS: -g2 -DPR_TUNABLE_XFER_LOG_MODE=0664 -DAIX7 -I/home/dan/proftpd-git -I/home/dan/proftpd-git/include -maix64 -Wall -fno-omit-frame-pointer
  LDFLAGS: -L$(top_srcdir)/lib -L$(top_builddir)/lib -dn -L/opt/freeware/lib -L/opt/freeware/lib64 -maix64 -Wl,-bbigtoc
  LIBS:  -lssl -lcrypto -lpam  -lcrypto -lsupp

  Files:
    Configuration File:
      /opt/proftpd/etc/proftpd.conf
    Pid File:
      /opt/proftpd/var/proftpd.pid
    Scoreboard File:
      /opt/proftpd/var/proftpd.scoreboard

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    - Autoshadow support
    - Controls support
    + curses support
    - Developer support
    - DSO support
    + IPv6 support
    + Largefile support
    - Lastlog support
    - Memcache support
    - ncurses support
    - NLS support
    - Redis support
    - Sodium support
    + OpenSSL support
    - PCRE support
    - POSIX ACL support
    - Shadow file suppport
    - Sendfile support
    + Trace support
    - xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 1024
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```
[proftpd.log](https://github.com/proftpd/proftpd/files/1953173/proftpd.log)
[trace.log](https://github.com/proftpd/proftpd/files/1953174/trace.log)
[proftpd_sftp.log](https://github.com/proftpd/proftpd/files/1953377/proftpd_sftp.log)

[proftpd.conf.txt](https://github.com/proftpd/proftpd/files/1953184/proftpd.conf.txt)

I've include a patch mod_auth.c that adds tracing that will show the timing of calls even when this is run on a Linux system:

[mod_auth.patch.txt](https://github.com/proftpd/proftpd/files/1953183/mod_auth.patch.txt)

**Comment by ws0w** [comment]
By the way, this problem does not happen on proftpd 1.3.5b.

**Comment by Castaglia** [comment]
Excellent find.  I suspect that this behavior is fallout from [Bug#4285](http://bugs.proftpd.org/show_bug.cgi?id=4285).

**Comment by ws0w** [comment]
Any progress on this?  Anything I can do to help?

**Comment by ws0w** [comment]
I was able to fix this bug by moving the loginsuccess() call to mod_auth_uninx.c in pw_check, right after a successful call to authenticate(). 

Perhaps all of the login_failed() and login_success() logic should be moved from mod_auth.c to mod_auth_unix.c where it can happen before PRIVS_RELINQUISH.

**Comment by ws0w** [comment]
I only fixed part of the problem. The call to loginsuccess() now does what it is supposed to do, but the unsuccessful_login_count is still being incremented incorrectly.

**Comment by Castaglia** [comment]
I don't have access to AIX for testing this.  However, I think I can create a local setup that reproduce this behavior, by hand-tweaking some of the build system to _think_ it's an AIX system -- and by providing stubs of the AIX functions.  Hoping to make progress on this soon.

**Comment by Castaglia** [comment]
I'll be addressing this through a series of PRs, trying to tackle the different edge cases/flows individually.  @ws0w will you be able to help me evaluate each one of the PRs?

**Comment by ws0w** [comment]
Yes, I should be able to help.

**Comment by Castaglia** [comment]
I guess a lot of the changes surround just _when_ `mod_sftp` would call `LOG_CMD_ERR` handlers for a "fake" PASS command (to indicate a failed login), considering the password, publickey, _et al_ authentication mechanisms that SSH provides.  And just how much of the issue, here, resolves around that `LOG_CMD_ERR` behavior _in conjunction with_ the use of AIX `loginfailed()`...

Given your experience with AIX, under what conditions would you _expect_ to see `loginfailed()` called (or under what conditions would you _not_ expect to see it called)?

**Comment by ws0w** [comment]
I would expect loginfailed() to be called after all authentication methods fail. Right now, it seems to be calling loginfailed() once when it can't find a public key, then again just before asking for a password. So, the failed login count is 2 before the first password prompt. SSH doesn't call it until after a non-empty bad password (or the second empty password).

This might not be a big problem if it was calling loginsuccess() after the authentication succeeds.

**Comment by Castaglia** [comment]
This got me to thinking.  The current implementation in `mod_sftp`, where we create fake "USER" and "PASS" commands and dispatch them, during the handling of SSH authentication requests, might be at the wrong layer.  That is, the success/failure of those fake commands should probably not be handled in the mechanism-specific code paths, but higher up.  I'll create a PR to show what I mean -- and it might just address this issue, too.

**Comment by Castaglia** [comment]
OK, @ws0w , can you evaluate these two PRs?
* https://github.com/proftpd/proftpd/pull/976 for changing when the command handlers are called
* https://github.com/proftpd/proftpd/pull/966 for calling `loginsuccess` while we still have root privs

**Comment by ws0w** [comment]
So far, I haven't been able to get configure to run cleanly. It's failing with:

checking size of short... configure: error: cannot compute sizeof (short)

I've hit this before on AIX, but don't remember what the fix is.

**Comment by Castaglia** [comment]
Hrm.  Usually the generated `config.log` file will show the checks, the code that `configure` tried to compile/run, the errors, _etc_.  Perhaps there are some clues there?

**Comment by ws0w** [comment]
Yeah, I found it ... there's a line in configure that adds "-rdynamic" to the LDFLAGS and the compiler doesn't recognize it. I'm back in business.

**Comment by ws0w** [comment]
It's working as expected now. A successful login resets the failed login count.

**Comment by Castaglia** [comment]
@ws0w for the sake of posterity, can you provide the tweaks/commands you used, for getting ProFTPD to compile on AIX?  I ask because I suspect these [ProFTPD AIX docs](https://github.com/proftpd/proftpd/blob/master/README.AIX) could use some updating...?

**Comment by Castaglia** [comment]
Excellent news that with both PRs, it's now behaving as you expect.  Does the success/failed login counts also behave as you expect for plain FTP logins, in addition to the tested SFTP logins?

**Comment by ws0w** [comment]
Yes, I tested both FTP and SFTP. Both work as expected.

I'm compiling with gcc 7.1.0

I edited the configure script to comment out lines 15611-15613.

Below is the configure command invocation that I used:

AR="ar -X64" LDFLAGS='-dn -L/opt/proftpd/lib -L/opt/freeware/lib -L/opt/freeware/lib64 -maix64 -Wl,-bbigtoc' CFLAGS='-DPR_TUNABLE_XFER_LOG_MODE=0664 -DAIX7 -I/home/dan/proftpd-test -I/home/dan/proftpd-test/include -maix64' ./configure \
   --prefix=/opt/proftpd \
   --enable-openssl \
   --enable-ctrls \
   --disable-auth-pam \
   --with-modules=mod_tls:mod_sftp \
   --disable-shared \
   --disable-dso \
   --enable-sendfile \
   --disable-strip \
   --enable-tunable-buffer-size

**Comment by Castaglia** [comment]
> I edited the configure script to comment out lines 15611-15613.

That part was added by https://github.com/proftpd/proftpd/commit/dae4cad8acce37174ad9f3cbcb30fa89439816e0; in https://github.com/proftpd/proftpd/commit/5122f072e3cfebf2de946715b8a5315ffafb557d I've now skipped the addition of `-rdynamic` on AIX7.  Let me know if I should expand that to cover other AIX versions.

**Comment by Castaglia** [comment]
Thanks for the AIX build instructions!  Looks like what we currently have might be dated, but still viable.  Otherwise, now, all fixes should be committed to `master` for this.

**Comment by Castaglia** [comment]
@ws0w Unfortunately I think I'm going to need to revert these changes; see Issue #1129.  This means that I'll try to work on a different approach to solve the original issues here.

**Comment by Castaglia** [comment]
Revisited fixes merged to master, and backported to the 1.3.7 branch.
